### PR TITLE
fix: [CDS-80902]: Add retries/errors for GitOps steps

### PR DESCRIPTION
**Pull request by meenaravichandran1**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54411)

### CODE REVIEW

File: product/ci/scm/file/file.go [hunk 1]
```diff
@@ -108,6 +113,23 @@ func FindFile(ctx context.Context, fileRequest *pb.GetFileRequest, log *zap.Suga
 	return out, nil
 }
 
+func findFileWithRetry(ctx context.Context, log *zap.SugaredLogger, client *scm.Client, fileRequest *pb.GetFileRequest, ref string, retryAttempts int) (content *scm.Content, response *scm.Response, err error) {
```
Comment (shivagowda) [inline product/ci/scm/file/file.go:116]: add unit test. Thanks!

**Comment by meenaravichandran1** [comment]
trigger uts

File: product/ci/scm/file/file.go (hunk 1 above)
Comment (meenaravichandran1) [reply product/ci/scm/file/file.go]: this change is updated

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ScmFetchFilesHelperNG.java [hunk 2]
```diff
@@ -240,7 +243,7 @@ private List<GitFile> fetchAnyFilesFromRepo(
     return gitFiles;
   }
 
-  private void throwFailedToFetchFileException(boolean useBranch, String branch, String commitId) {
+  private void throwFailedToFetchConnectivityError(boolean useBranch, String branch, String commitId) {
```
[the remark this answers is not in the document]
Comment (meenaravichandran1) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ScmFetchFilesHelperNG.java:246]: this error is based on network connectivity and connectivityError seemed more appropriate to me than connectionException. Will think of a better name. 

Let me know if you have any suggestions too.

File: 960-api-services/src/main/java/io/harness/cistatus/service/gitlab/GitlabServiceImpl.java [hunk 3]
```diff
@@ -67,11 +67,13 @@ public JSONObject mergePR(String apiUrl, String slug, String token, String prNum
           continue;
         }
         // Possible status codes https://docs.gitlab.com/ee/api/merge_requests.html#merge-a-merge-request
-        if (response.code() != 405) {
+        // 422 could happen intermittently in Gitlab
+        // https://platform.harness.internal/ng/account/W50osoQJS42JItbue3ddhA/cd/orgs/default/projects/meenatest/pipelines/test/deployments/xVWJQ9D0RTeIzVgmtixYXA/pipeline?storeType=INLINE
+        if (response.code() != 405 || response.code() != 422) {
```
Comment (manavjot-harness) [inline 960-api-services/src/main/java/io/harness/cistatus/service/gitlab/GitlabServiceImpl.java:72]: this should be &&

Comment (meenaravichandran1) [reply 960-api-services/src/main/java/io/harness/cistatus/service/gitlab/GitlabServiceImpl.java]: done

**Comment by meenaravichandran1** [comment]
trigger uts

**Comment by meenaravichandran1** [comment]
trigger uts

**Comment by shivagowda** [comment]
trigger review

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ScmFetchFilesHelperNG.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (manavjot-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ScmFetchFilesHelperNG.java:246]: yea maybe name it to throwFailedToFetchConnectivityException?

[the remark this answers is not in the document]
Comment (meenaravichandran1) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ScmFetchFilesHelperNG.java]: Renamed it to throwFailedToFetchNoConnectivityException

**Comment by meenaravichandran1** [comment]
trigger uts

**Review by manavjot-harness — APPROVED** [review]

**Review by mankrit-singh — APPROVED** [review]

**Comment by meenaravichandran1** [comment]
trigger review

**Pull request merged**